File: app/src/test/java/com/abishov/hexocat/github/RepositoryIntegrationTests.java
package com.abishov.hexocat.github;

import com.abishov.hexocat.Inject;
import com.google.gson.Gson;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import static org.assertj.core.api.Java6Assertions.assertThat;

@RunWith(JUnit4.class)
public class RepositoryIntegrationTests {

    @Test
    public void payloadMustMapToModelCorrectly() {
        Gson gson = Inject.gson();
        Repository repository = gson.fromJson("{\n" +
                "      \"id\": 892275,\n" +
                "      \"name\": \"retrofit\",\n" +
                "      \"forks\": 222," +
                "      \"stargazers_count\": 777," +
                "      \"html_url\": \"https://github.com/square/retrofit\",\n" +
                "      \"description\": \"Type-safe HTTP client for Android and Java by Square, Inc.\"," +
                "      \"owner\": {\n" +
                "        \"login\": \"square\",\n" +
                "        \"id\": 82592,\n" +
                "        \"avatar_url\": \"https://avatars0.githubusercontent.com/u/82592?v=4\",\n" +
                "        \"gravatar_id\": \"\",\n" +
                "        \"url\": \"https://api.github.com/users/square\",\n" +
                "        \"html_url\": \"https://github.com/square\",\n" +
                "        \"followers_url\": \"https://api.github.com/users/square/followers\",\n" +
                "        \"following_url\": \"https://api.github.com/users/square/following{/other_user}\",\n" +
                "        \"gists_url\": \"https://api.github.com/users/square/gists{/gist_id}\",\n" +
                "        \"starred_url\": \"https://api.github.com/users/square/starred{/owner}{/repo}\",\n" +
                "        \"subscriptions_url\": \"https://api.github.com/users/square/subscriptions\",\n" +
                "        \"organizations_url\": \"https://api.github.com/users/square/orgs\",\n" +
                "        \"repos_url\": \"https://api.github.com/users/square/repos\",\n" +
                "        \"events_url\": \"https://api.github.com/users/square/events{/privacy}\",\n" +
                "        \"received_events_url\": \"https://api.github.com/users/square/received_events\",\n" +
                "        \"type\": \"User\",\n" +
                "        \"site_admin\": false\n" +
                "      }" +
                "}", Repository.class);

        assertThat(repository.name()).isEqualTo("retrofit");
        assertThat(repository.forks()).isEqualTo(222);
        assertThat(repository.stars()).isEqualTo(777);
        assertThat(repository.htmlUrl()).isEqualTo("https://github.com/square/retrofit");
        assertThat(repository.description()).isEqualTo("Type-safe HTTP client for Android and Java by Square, Inc.");
        assertThat(repository.owner().login()).isEqualTo("square");
        assertThat(repository.owner().htmlUrl()).isEqualTo("https://github.com/square");
    }
}

File: app/src/test/java/com/abishov/hexocat/github/RepositoryIntegrationTests.kt
package com.abishov.hexocat.github

import assertk.assert
import assertk.assertions.isEqualTo
import com.abishov.hexocat.Inject
import org.junit.Test

class RepositoryIntegrationTests {

  @Test
  fun `payload must map to model correctly`() {
    val adapter = Inject.moshi()
      .adapter(Repository::class.java)

    val repository = adapter.fromJson(
      """
        {
          "id": 892275,
          "name": "retrofit",
          "forks": 222,
          "stargazers_count": 777,
          "html_url": "https://github.com/square/retrofit",
          "description": "Type-safe HTTP client",
          "owner": {
            "login": "square",
            "id": 82592,
            "avatar_url": "https://avatars0.githubusercontent.com/u/82592?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/square",
            "html_url": "https://github.com/square",
            "followers_url": "https://api.github.com/users/square/followers",
            "following_url": "https://api.github.com/users/square/following{/other_user}",
            "gists_url": "https://api.github.com/users/square/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/square/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/square/subscriptions",
            "organizations_url": "https://api.github.com/users/square/orgs",
            "repos_url": "https://api.github.com/users/square/repos",
            "events_url": "https://api.github.com/users/square/events{/privacy}",
            "received_events_url": "https://api.github.com/users/square/received_events",
            "type": "User",
            "site_admin": false
          }
        }
        """
    )!!

    assert(repository.name).isEqualTo("retrofit")
    assert(repository.forks).isEqualTo(222)
    assert(repository.stars).isEqualTo(777)
    assert(repository.htmlUrl).isEqualTo("https://github.com/square/retrofit")
    assert(repository.description).isEqualTo("Type-safe HTTP client")
    assert(repository.owner.login).isEqualTo("square")
    assert(repository.owner.htmlUrl).isEqualTo("https://github.com/square")
  }
}
